src/log-query/src/log_query.rs

```diff
@@ -0,0 +1,322 @@
+                return Err(InvalidTimeFilterSnafu {
+                    filter: self.clone(),
+                }
+                .build());
+            }
+            start_dt = Some(start);
+            end_dt = end_opt;
+        } else if self.start.is_some() && self.end.is_some() {
+            // Both 'start' and 'end' are provided
+            let (start, _) = Self::parse_datetime(self.start.as_ref().unwrap())?;
+            let (end, _) = Self::parse_datetime(self.end.as_ref().unwrap())?;
+            start_dt = Some(start);
+            end_dt = Some(end);
+        } else if self.span.is_some() && (self.start.is_some() || self.end.is_some()) {
+            // 'span' with 'start' or 'end'
+            let span = Self::parse_span(self.span.as_ref().unwrap())?;
+            if self.start.is_some() {
+                let (start, _) = Self::parse_datetime(self.start.as_ref().unwrap())?;
+                let end = start + span;
+                start_dt = Some(start);
+                end_dt = Some(end);
+            } else {
+                let (end, _) = Self::parse_datetime(self.end.as_ref().unwrap())?;
+                let start = end - span;
+                start_dt = Some(start);
+                end_dt = Some(end);
+            }
+        } else if self.span.is_some() && self.start.is_none() && self.end.is_none() {
+            // Only 'span' is provided
+            let span = Self::parse_span(self.span.as_ref().unwrap())?;
+            let end = Utc::now();
+            let start = end - span;
+            start_dt = Some(start);
+            end_dt = Some(end);
+        } else if self.start.is_some() && self.span.is_some() && self.end.is_some() {
+            // All fields are provided; 'start' and 'end' take priority
+            let (start, _) = Self::parse_datetime(self.start.as_ref().unwrap())?;
+            let (end, _) = Self::parse_datetime(self.end.as_ref().unwrap())?;
+            start_dt = Some(start);
+            end_dt = Some(end);
+        } else {
+            // Exception
+            return Err(InvalidTimeFilterSnafu {
+                filter: self.clone(),
+            }
+            .build());
+        }
+
+        // Validate that end is after start
+        if let (Some(start), Some(end)) = (&start_dt, &end_dt) {
+            if end <= start {
+                return Err(EndBeforeStartSnafu {
+                    start: start.to_rfc3339(),
+                    end: end.to_rfc3339(),
+                }
+                .build());
+            }
+        }
+
+        // Update the fields with canonicalized timestamps
+        if let Some(start) = start_dt {
+            self.start = Some(start.to_rfc3339());
+        }
+
+        if let Some(end) = end_dt {
+            self.end = Some(end.to_rfc3339());
+        }
+
+        Ok(())
+    }
+
+    /// Util function returns a start and optional end DateTime
+    fn parse_datetime(s: &str) -> Result<(DateTime<Utc>, Option<DateTime<Utc>>)> {
```

Comment:
It looks like a very generic parse function, should it be moved to `common-time`?
Same as `parse_span` below.

Reply:
`parse_datetime` is not that generic for now as here is the first and only place that accepts both date and timestamp. I'd do a refactor afterward for `parse_span` to combine the same piece of logic in the plan rewriter.